Don't check BC/FC for ops with valid upgrader

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72193

This PR removes the operators that have valid upgraders from the BC table. 

Differential Revision: [D33948943](https://our.internmc.facebook.com/intern/diff/D33948943)

tugsbayasgalan committed:
Don't check BC/FC for ops with valid upgrader

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33948943).

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -155,6 +153,32 @@ def load_schemas_to_dict():
+        return False
+
+    current_version = torch._C._get_max_operator_version()
+    for i in possible_overloads:
```
Nit:
rename `i` to something like `schema_name`. `i` gives the impression that it is a number

gmagogsfm commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1756,6 +1756,7 @@ void initJitScriptBindings(PyObject* module) {
   m.def("_get_max_operator_version", &getMaxOperatorVersion);
   m.def("_get_operator_version_map", &get_operator_version_map);
   m.def("_get_upgrader_ranges", &getUpgradersRangeForOp);
+  m.def("_is_op_current", &isOpSymbolCurrent);
```
Normally, it is slightly better for C++ function and Python method name to approximately match to reduce cognitive load when reading code.

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -155,6 +153,32 @@ def load_schemas_to_dict():
+            possible_overloads.add(key)
+
+    # there is no entry for this schema
+    if len(possible_overloads) == 0:
```
Do we need to verify that these upgraders actually have same schema (not just same op name) as `existing_schema`?

tugsbayasgalan commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -155,6 +153,32 @@ def load_schemas_to_dict():
     return new_schema_dict
 
+def has_valid_upgraders(schema):
+    version_map = torch._C._get_operator_version_map()
```
It is hard because if we split the key and look up by op name, there will be duplicates because of overloads.

gmagogsfm replied:
You can process the map to be [op_name, pair< overload, schema>] or whatever that's convenient for look up.

tugsbayasgalan commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -155,6 +153,32 @@ def load_schemas_to_dict():
+    if len(possible_overloads) == 0:
+        return False
+
+    current_version = torch._C._get_max_operator_version()
```
Yeah sounds good. This existed before this PR

tugsbayasgalan committed:
Update on "Don't check BC/FC for ops with valid upgrader"


This PR removes the operators that have valid upgraders from the BC table. 

Differential Revision: [D33948943](https://our.internmc.facebook.com/intern/diff/D33948943)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33948943).

mruberry commented:
Can you elaborate on this PR, @tugsbayasgalan? Like just because an op has an upgrader doesn't mean that future BC-breaking changes can't be made to it, or that users won't experience a BC-breaking change.

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -155,14 +35,53 @@ def load_schemas_to_dict():
         new_schema_dict[s.name].append(s)
     return new_schema_dict
 
+def has_valid_upgraders(schema, version_map):
```
We don't require upgraders for every BC-breaking change that would trigger this test today. For example, ops in prototype and beta modules have no BC guarantees.
